### DISCUSSION TITLE
AbstractSemanticTokensProvider.getDocumentTokensBuilder

### DISCUSSION THREAD

**Discussion by cdietrich** [eclipse-langium/langium#1283] [Q&A]
Hi, is there a reason `AbstractSemanticTokensProvider.getDocumentTokensBuilder`
reuses
SemanticTokensBuilder
i get this warning on the client side on the 2nd call  `Overlapping semantic tokens detected at lineNumber 1, column 10`

<img width="1281" alt="grafik" src="https://github.com/eclipse-langium/langium/assets/238736/49ae3b89-6a0a-4be4-8a3b-079ee261d7ac">

here in _data there is already the previous tokens from the previous call.
is there something missing that flushes? e.g previousResult call?

pot. workaround:

```
    protected override getDocumentTokensBuilder(document: LangiumDocument): SemanticTokensBuilder {
        const existing = this.tokensBuilders.get(document.uri.toString());
        if (existing) {
            existing.previousResult(existing.id)
            return existing;
        }
        const builder = new SemanticTokensBuilder();
        this.tokensBuilders.set(document.uri.toString(), builder);
        return builder;
    }
    ```

**Comment by msujew** [comment]
We do compute a delta based on the previous result for each open file. See [here](https://github.com/eclipse-langium/langium/blob/7066ac43e575d5c032e30e32bab67e5895d5f40a/packages/langium/src/lsp/semantic-token-provider.ts#L220-L227). 

I think using `existing.previousResult(existing.id)` for every request flushes the whole cache, which makes delta updates kind of useless. But we should probably use it for every request that isn't a delta request.

**Reply by cdietrich** [reply]
hmm then the question is why it is not working.
=> should we flush at the begin  of highlighting and then build delta?

**Reply by cdietrich** [reply]
=> should `semanticHighlight` and `semanticHighlightRange` flush?

**Reply by msujew** [reply]
Yes, that's what I meant. They should probably flush but currently don't do that.

**Reply by msujew** [reply]
Can you check whether these overrides also solve your issue? So far, I'm having trouble reproducing this.

```ts
    async semanticHighlight(document: LangiumDocument, _params: SemanticTokensParams, cancelToken = CancellationToken.None): Promise<SemanticTokens> {
        this.currentRange = undefined;
        this.currentDocument = document;
        this.currentTokensBuilder = this.getDocumentTokensBuilder(document);
        this.currentTokensBuilder.previousResult(this.currentTokensBuilder.id);
        await this.computeHighlighting(document, this.createAcceptor(), cancelToken);
        return this.currentTokensBuilder.build();
    }

    async semanticHighlightRange(document: LangiumDocument, params: SemanticTokensRangeParams, cancelToken = CancellationToken.None): Promise<SemanticTokens> {
        this.currentRange = params.range;
        this.currentDocument = document;
        this.currentTokensBuilder = this.getDocumentTokensBuilder(document);
        this.currentTokensBuilder.previousResult(this.currentTokensBuilder.id);
        await this.computeHighlighting(document, this.createAcceptor(), cancelToken);
        return this.currentTokensBuilder.build();
    }
```

**Reply by cdietrich** [reply]
yes it does

**Reply by msujew** [reply]
Alright, that's good. Then I'll create a PR for that. It definitely sounds reasonable to call `previousResult` with the existing ID (thereby flushing the cache), in case we don't have a delta update. The code for the `SemanticTokenBuilder` on Microsoft's side is pretty undocumented, so it wasn't quite clear on how to use it...
